cli: TestHalfOnlineLossOfQuorumRecovery failed

cli.TestHalfOnlineLossOfQuorumRecovery [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21077416?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21077416?buildTab=artifacts#/) on master @ [47eae13ca072761c1708116e66776ca01f8d96f9](https://github.com/cockroachdb/cockroach/commits/47eae13ca072761c1708116e66776ca01f8d96f9):


```
=== RUN   TestHalfOnlineLossOfQuorumRecovery
    test_log_scope.go:172: test logs captured to: /artifacts/tmp/_tmp/6c689cc523db9d8596d03c0a6ef213e4/logTestHalfOnlineLossOfQuorumRecovery3776359264
    test_log_scope.go:82: use -show-logs to present logs inline
[debug recover make-plan --confirm=y --certs-dir=test_certs --host=127.0.0.1:36465 --cluster-name=TestCluster-uhPiKE1Ptl --plan=/artifacts/tmp/_tmp/6c689cc523db9d8596d03c0a6ef213e4/TestHalfOnlineLossOfQuorumRecovery3559714859/recovery-plan.json]
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x42ee854]

goroutine 17444 [running]:
github.com/cockroachdb/cockroach/pkg/kv/kvserver/loqrecovery.Server.ServeClusterReplicas.serveClusterReplicasParallelFn.func3(0x3, {0x7453ed0, 0x4009352bf8})
	pkg/kv/kvserver/loqrecovery/server.go:268 +0x1b4
github.com/cockroachdb/cockroach/pkg/kv/kvserver/loqrecovery.visitNodeWithRetry({0x7397aa8, _}, _, {_, _, _, _, _, _, _, ...}, ...)
	pkg/kv/kvserver/loqrecovery/server.go:786 +0x1e8
github.com/cockroachdb/cockroach/pkg/kv/kvserver/loqrecovery.NewServer.makeVisitAvailableNodesInParallel.func1.3()
	pkg/kv/kvserver/loqrecovery/server.go:752 +0x70
golang.org/x/sync/errgroup.(*Group).Go.func1()
	external/org_golang_x_sync/errgroup/errgroup.go:93 +0x4c
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 17425
	external/org_golang_x_sync/errgroup/errgroup.go:78 +0x90
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestHalfOnlineLossOfQuorumRecovery.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-59556

--- a/pkg/cli/debug_recover_loss_of_quorum_test.go
+++ b/pkg/cli/debug_recover_loss_of_quorum_test.go
@@ -486,7 +486,7 @@ func TestHalfOnlineLossOfQuorumRecovery(t *testing.T) {
 				},
 				LOQRecovery: &loqrecovery.TestingKnobs{
 					MetadataScanTimeout: 15 * time.Second,
-					ForwardReplicaFilter: func(
+					MaybeInjectError: func(
 						response *serverpb.RecoveryCollectLocalReplicaInfoResponse,
 					) error {
 						// Artificially add an error that would cause the server to retry

--- a/pkg/kv/kvserver/loqrecovery/server.go
+++ b/pkg/kv/kvserver/loqrecovery/server.go
@@ -77,7 +77,7 @@ type Server struct {
 	decommissionFn     func(context.Context, roachpb.NodeID) error
 
 	metadataQueryTimeout time.Duration
-	forwardReplicaFilter func(*serverpb.RecoveryCollectLocalReplicaInfoResponse) error
+	maybeInjectError     func(*serverpb.RecoveryCollectLocalReplicaInfoResponse) error
 }
 
 func NewServer(
@@ -95,12 +95,12 @@ func NewServer(
 	// effort operations where cluster info collection as an operation succeeds
 	// even if some parts of it time out.
 	metadataQueryTimeout := 1 * time.Minute
-	var forwardReplicaFilter func(*serverpb.RecoveryCollectLocalReplicaInfoResponse) error
+	var maybeInjectError func(*serverpb.RecoveryCollectLocalReplicaInfoResponse) error
 	if rk, ok := knobs.(*TestingKnobs); ok {
 		if rk.MetadataScanTimeout > 0 {
 			metadataQueryTimeout = rk.MetadataScanTimeout
 		}
-		forwardReplicaFilter = rk.ForwardReplicaFilter
+		maybeInjectError = rk.MaybeInjectError
 	}
 	return &Server{
 		nodeIDContainer:      nodeIDContainer,
@@ -112,7 +112,7 @@ func NewServer(
 		planStore:            planStore,
 		decommissionFn:       decommission,
 		metadataQueryTimeout: metadataQueryTimeout,
-		forwardReplicaFilter: forwardReplicaFilter,
+		maybeInjectError:     maybeInjectError,
 	}
 }
 
@@ -221,13 +221,13 @@ func (s Server) ServeClusterReplicas(
 		fanOutConnectionRetryOptions,
 		req.MaxConcurrency,
 		allNodes,
-		serveClusterReplicasParallelFn(ctx, syncOutStream, s.forwardReplicaFilter, &replicas, &nodes))
+		serveClusterReplicasParallelFn(ctx, syncOutStream, s.maybeInjectError, &replicas, &nodes))
 }
 
 func serveClusterReplicasParallelFn(
 	ctx context.Context,
 	outStream *threadSafeStream[*serverpb.RecoveryCollectReplicaInfoResponse],
-	forwardReplicaFilter func(*serverpb.RecoveryCollectLocalReplicaInfoResponse) error,
+	maybeInjectError func(*serverpb.RecoveryCollectLocalReplicaInfoResponse) error,
 	replicas, nodes *atomic.Int64,
 ) visitNodeAdminFn {
 	return func(nodeID roachpb.NodeID, client serverpb.RPCAdminClient) error {
@@ -244,8 +244,8 @@ func serveClusterReplicasParallelFn(
 			if err == io.EOF {
 				break
 			}
-			if forwardReplicaFilter != nil {
-				err = forwardReplicaFilter(r)
+			if err == nil && maybeInjectError != nil {
+				err = maybeInjectError(r)
 			}
 			if err != nil {
 				// Some replicas were already sent back, need to notify client of stream

--- a/pkg/kv/kvserver/loqrecovery/server_integration_test.go
+++ b/pkg/kv/kvserver/loqrecovery/server_integration_test.go
@@ -143,7 +143,7 @@ func TestStreamRestart(t *testing.T) {
 			Knobs: base.TestingKnobs{
 				LOQRecovery: &loqrecovery.TestingKnobs{
 					MetadataScanTimeout: 15 * time.Second,
-					ForwardReplicaFilter: func(response *serverpb.RecoveryCollectLocalReplicaInfoResponse) error {
+					MaybeInjectError: func(response *serverpb.RecoveryCollectLocalReplicaInfoResponse) error {
 						if response.ReplicaInfo.NodeID == 2 && response.ReplicaInfo.Desc.RangeID == 14 && failCount.Add(1) < 3 {
 							return errors.New("rpc stream stopped")
 						}

--- a/pkg/kv/kvserver/loqrecovery/testing_knobs.go
+++ b/pkg/kv/kvserver/loqrecovery/testing_knobs.go
@@ -17,9 +17,10 @@ type TestingKnobs struct {
 	// behaviors when meta ranges are unavailable.
 	MetadataScanTimeout time.Duration
 
-	// Replica filter for forwarded replica info when collecting fan-out data.
-	// It can be called concurrently, so must be safe for concurrent use.
-	ForwardReplicaFilter func(*serverpb.RecoveryCollectLocalReplicaInfoResponse) error
+	// MaybeInjectError can be used to inject an error after replica info is
+	// received when collecting fan-out data. It can be called concurrently,
+	// so must be safe for concurrent use.
+	MaybeInjectError func(*serverpb.RecoveryCollectLocalReplicaInfoResponse) error
 }
 
 // ModuleTestingKnobs is part of the base.ModuleTestingKnobs interface.

